### ISSUE TITLE
LightningModule.log with no reduce.

### ISSUE THREAD

**Issue by dreaquil** [PyTorchLightning/pytorch-lightning#12497]
## 🚀 Feature

I would like to be able to log a roc metric without reducing (mean etc.)

### Motivation

I think it would be useful to be able to return the ROC curve parameters (fprs, tprs, thresholds) so that the threshold can be determined by some other function.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @awaelchli @rohitgr7 @akihironitta @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

**Comment by rohitgr7** [comment]
hey @dreaquil !

if you are using torchmetrics, you can collect the states as shown in the example and return them to be able to compute them as per your requirement and then log it.
https://torchmetrics.readthedocs.io/en/latest/references/modules.html#roc

**Comment by dreaquil** [comment]
Hi @rohitgr7,

Thanks for your reply! The issue is that when you call self.log from a lightningmodule, it is not possible to log the result from the roc (FPR, TPR, thresholds) without reducing. FPR, TPR and thresholds are lists which should not be averaged.

**Comment by rohitgr7** [comment]
can you share pseudocode on how you want this to work out for your use case?

**Comment by dreaquil** [comment]
Hi @rohitgr7!

```
>>> pred = torch.tensor([[0.8191, 0.3680, 0.1138],
...                      [0.3584, 0.7576, 0.1183],
...                      [0.2286, 0.3468, 0.1338],
...                      [0.8603, 0.0745, 0.1837]])
>>> target = torch.tensor([[1, 1, 0], [0, 1, 0], [0, 0, 0], [0, 1, 1]])
>>> roc = ROC(num_classes=3, pos_label=1)
>>> fpr, tpr, thresholds = roc(pred, target)
>>> fpr
tensor([0., 0., 0., 0., 1.])
>>> my_lightning_module.log('fpr', fpr, reduce_fx=None)
```
Following logging, I would like the fpr to appear in the `my_lightning_module.callback_metrics` as a whole as follows:
`{'fpr': tensor([0., 0., 0., 0., 1.])}`

Currently, since it is not possible to set `reduce_fx=None`, it defaults to `reduce_fx='mean'`. The resultant `my_lightning_module.callback_metrics` is as follows:

`{'fpr': tensor(0.2)}`

**Comment by rohitgr7** [comment]
that is not possible since logging only supports scalar metrics to be logged.
check out the description: https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#logging-from-a-lightningmodule

> I think it would be useful to be able to return the ROC curve parameters (fprs, tprs, thresholds) so that the threshold can be determined by some other function.

you can use the states from the ROC instance in your `validation_epoch_end` and compute the scalar metric accordingly using those states :)

**Comment by dreaquil** [comment]
Might it make sense to start supporting lists of scalar values or is there some technical reason this is not doable?

**Comment by rohitgr7** [comment]
> Might it make sense to start supporting lists of scalar values or is there some technical reason this is not doable?

yes, `self.log` aggregates scalars, and logs them to the progress bar and to the loggers. Not all progress bar and loggers support logging non-scalars and aggregation with `on_epoch=True` is not possible for non-scalars.

**Comment by dreaquil** [comment]
@rohitgr7 Thanks for the clarification!
